source/task_scheduler.c

```diff
@@ -122,6 +124,16 @@ int aws_task_scheduler_schedule_future(struct aws_task_scheduler *scheduler,
 
     struct task_container container;
 
+    /* this serves the purpose of handling reentrant scheduling while the tasks are being run.
+       if the clock tick is on the same nanosecond (with microsecond precision) as the run was kicked
+       off, it will still be run. As a result, if time_to_run is before or on the same clock tick, increment it
+       by one nanosecond to avoid the issue. However this does assume that the clock ticks are at least greater than 0. Seeing as how
+       boot times take longer than 1 nanosecond and a system clock will always be greater than 0, I think this is a safe assumption 
+       in a real world scenario. */
+    if (AWS_UNLIKELY(time_to_run <= scheduler->latest_run_time)) {
+        time_to_run = scheduler->latest_run_time + 1;
```

Comment:
Later in this review you had to change a unit test to start its clock at 1 instead of 0.

If this variable was `scheduler->run_time_min` instead of `scheduler->latest_run_time`, then we could allow tasks at time 0.